### PR TITLE
pkg/types/config: Break into pkg/validate and pkg/tfvars

### PR DESCRIPTION
**Pull request by wking**
Builds on #403; I'll rebase after that lands.

This removes a few `Cluster` properties which are no longer used by the new installer.  It also drops YAML loading for Cluster now that it's a one-way map from `InstallConfig` to Terraform JSON.  I've shifted validation into its own package, since we should be validating these at the `InstallConfig` level.

I've dropped the old `Cluster` (and subtype) specific validation.  We'll want to add `InstallConfig` validation once we support loading it from the disk, but for now we can rely on the `UserProvided` validation (and this commit was already large enough ;).

### CODE REVIEW

File: pkg/types/config/cluster.go [hunk 1]
```diff
@@ -28,94 +25,29 @@ const (
...
-		Type:        tectonicnetwork.NetworkFlannel,
-	},
-	OpenStack: openstack.OpenStack{
-		Region:           openstack.DefaultRegion,
-		NetworkCIDRBlock: openstack.DefaultNetworkCIDRBlock,
-	},
-}
-
 // Cluster defines the config for a cluster.
 type Cluster struct {
```
Comment (abhinavdahiya) [inline pkg/types/config/cluster.go]: Can we just rename this struct to `TFVars` as that's what this is. can be separate PR; but we are a lot about this struct so I thought name might also be in the area :P.

Comment (wking) [reply pkg/types/config/cluster.go]: > Can we just rename this struct to `TFVars`...

I've pushed 5882b2e -> 4d47856 to move this out of `pkg/types` too and simplify this package's [current API][1] to a single function:

```console
$ godoc ./pkg/tfvars/
PACKAGE DOCUMENTATION

package tfvars
    import "./pkg/tfvars/"

    Package tfvars converts an InstallConfig to Terraform variables.

FUNCTIONS

func TFVars(cfg *types.InstallConfig, bootstrapIgn string, masterIgns []string, workerIgn string) ([]byte, error)
    TFVars converts the InstallConfig and Ignition content to
    terraform.tfvar JSON.

SUBDIRECTORIES

	aws
	libvirt
	openstack

```

[1]: https://godoc.org/github.com/openshift/installer/pkg/types/config

File: pkg/asset/ignition/bootstrap/bootstrap.go [hunk 2]
```diff
@@ -190,7 +190,7 @@ func (a *bootstrap) getTemplateData(installConfig *types.InstallConfig) (*bootst
 	}
 	etcdEndpoints := make([]string, installConfig.MasterCount())
 	for i := range etcdEndpoints {
-		etcdEndpoints[i] = fmt.Sprintf("https://%s-etcd-%d.%s:2379", installConfig.Name, i, installConfig.BaseDomain)
+		etcdEndpoints[i] = fmt.Sprintf("https://%s-etcd-%d.%s:2379", installConfig.ObjectMeta.Name, i, installConfig.BaseDomain)
```
Comment (abhinavdahiya) [inline pkg/asset/ignition/bootstrap/bootstrap.go]: > Stop relying on field promotion [1], because I'm about to add
InstallConfig.Platform.Name().

No sure how they coorelates. I think we can keep is as is.
When in future when we split installconfig to installconfig and clusterconfig; we can change to clustername

**Comment by abhinavdahiya** [comment]
Can we not create a new package that has types. we already have pkg/types. maybe just a new file.

**Comment by wking** [comment]
> Can we not create a new package that has types. we already have `pkg/types`.

We already have types all over:

```console
$ git describe
v0.1.0-30-ge4c8898
$ for DIR in pkg/*; do echo -n "${DIR} "; git grep '^type ' "${DIR}" | wc -l; done
pkg/asset 72
pkg/destroy 5
pkg/ipnet 1
pkg/rhcos 0
pkg/terraform 1
pkg/types 50
```

My impression was that `pkg/types` was going to be for external consumers who wanted to use something we were pushing into the cluster (e.g. `InstallConfig`).  And that ideally, `pkg/types` would have very minimal external dependencies or dynamic code outside of the type definitions themselves.  This roughly corresponds to [our current docs for the package][1].

The new `pkg/tfvars` has an internal type, but its public API is just [the one function][2], which nobody except for us is likely to care about.  If my understanding of `pkg/types` scoping is accurate, then the `TFVars` function does not qualify for `pkg/types` inclusion.

[1]: https://github.com/openshift/installer/blob/e4c8898cc69921aee21f96a5c90e50fc461b5d02/pkg/types/doc.go#L1
[2]: https://github.com/openshift/installer/pull/412#discussion_r222477012

File: pkg/asset/ignition/bootstrap/bootstrap.go (hunk 2 above)
Comment (wking) [reply pkg/asset/ignition/bootstrap/bootstrap.go]: > > Stop relying on field promotion [1], because I'm about to add InstallConfig.Platform.Name().
>
> No sure how they coorelates. I think we can keep is as is.

```console
$ git describe --dirty
v0.1.0-11-ga0a1dbc-dirty
$ git diff
diff --git a/pkg/asset/ignition/bootstrap/bootstrap.go b/pkg/asset/ignition/bootstrap/bootstrap.go
index a5226aa..51e6fc9 100644
--- a/pkg/asset/ignition/bootstrap/bootstrap.go
+++ b/pkg/asset/ignition/bootstrap/bootstrap.go
@@ -190,7 +190,7 @@ func (a *bootstrap) getTemplateData(installConfig *types.InstallConfig) (*bootst
 	}
 	etcdEndpoints := make([]string, installConfig.MasterCount())
 	for i := range etcdEndpoints {
-		etcdEndpoints[i] = fmt.Sprintf("https://%s-etcd-%d.%s:2379", installConfig.ObjectMeta.Name, i, installConfig.BaseDomain)
+		etcdEndpoints[i] = fmt.Sprintf("https://%s-etcd-%d.%s:2379", installConfig.Name, i, installConfig.BaseDomain)
 	}
 
 	releaseImage := defaultReleaseImage
$ hack/build.sh 
++ dirname hack/build.sh
+ cd hack/..
+ MODE=release
++ git describe --always --abbrev=0 --dirty
+ LDFLAGS=' -X main.version=v0.1.0-dirty'
+ TAGS=
+ export CGO_ENABLED=0
+ CGO_ENABLED=0
+ case "${MODE}" in
+ TAGS=' release'
++ go env GOPATH
+ GOPATH=/home/trking/.local/lib/go/src/github.com/openshift/installer/vendor:/home/trking/.local/lib/go
+ go generate ./data
writing assets_vfsdata.go
+ echo ' release'
+ grep -q libvirt_destroy
+ go build -ldflags ' -X main.version=v0.1.0-dirty' -tags ' release' -o ./bin/openshift-install ./cmd/openshift-install
# github.com/openshift/installer/pkg/asset/ignition/bootstrap
pkg/asset/ignition/bootstrap/bootstrap.go:193:77: ambiguous selector installConfig.Name
```

So this is not compiling (with `ambiguous selector installConfig.Name`) unless I make this change.

Comment (abhinavdahiya) [reply pkg/asset/ignition/bootstrap/bootstrap.go]: Sounds good.

**Comment by wking** [comment]
Rebased onto master and green.
